### ISSUE TITLE
wrong server shutdown order?

### ISSUE THREAD

**Issue by novoxd** [fafhrd91/aiohttp#8179]
### Describe the bug

tldr example with aiojobs background task
```py
import asyncio
import time

from aiohttp import web
from aiojobs.aiohttp import setup, spawn

import logging
logging.basicConfig(level=logging.DEBUG)

async def back_task():
    st = time.time()
    while True:
        print(f'I am running for {int(time.time() - st)} seconds')
        try:
            await asyncio.sleep(1)
        except asyncio.CancelledError:
            print(f'After {int(time.time() - st)} seconds finished by asyncio cancel')
            return

async def handler(request):
    await spawn(request, back_task())
    return web.Response(text="ok")

aiohttp_app = web.Application()
aiohttp_app.router.add_get('/', handler)
setup(aiohttp_app)

web.run_app(aiohttp_app)
```
running and after 3 seconds from request stopping the server 
output:
```
DEBUG:asyncio:Using selector: EpollSelector
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
INFO:aiohttp.access:::1 [22/Feb/2024:10:53:47 +0300] "GET / HTTP/1.1" 200 153 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
I am running for 0 seconds
I am running for 1 seconds
I am running for 2 seconds
I am running for 3 seconds
^C
I am running for 4 seconds
I am running for 5 seconds
I am running for 6 seconds
...
I am running for 60 seconds
I am running for 61 seconds
I am running for 62 seconds
I am running for 63 seconds
After 63 seconds finished by asyncio cancel

Process finished with exit code 0
```

This happens because aiojobs [creates](https://github.com/aio-libs/aiojobs/blob/master/aiojobs/_scheduler.py#L42) monitoring background task while `cleanup_context` . What will be stopped on [68](https://github.com/aio-libs/aiojobs/blob/master/aiojobs/aiohttp.py#L68). But aiohttp server calls it only after 60sec timeout. This happens because [here](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web.py#L343) aiojobs bg task will be picked up for waiting on [311](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web.py#L311) what cannot be done because aiojobs bg task never receives cancel because `BaseRunner` `shutdown_callback` [runs](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_runner.py#L306) before `_server.shutdown`.

Changing [BaseRunner.cleanup](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_runner.py#L301) order to:
```py
if self._server:  # If setup succeeded
    self._server.pre_shutdown()
    await self.shutdown()

    await self._server.shutdown(self._shutdown_timeout)
await self._cleanup_server()

if self._server:
    if self.shutdown_callback:
        await self.shutdown_callback()
```
fixes the issue.

I am not sure am I right with the fix. It is quite strange that so long living project like aiohttp has this type of bug. Am I doing something wrong? I think it can be related to #5672 .

@Dreamsorcerer I know that you was around this. And @asvetlov is the right folk to question.

### To Reproduce

.

### Expected behavior

Background task finishes on server shutdown immediately.

### Logs/tracebacks

```python-traceback
.
```

### Python Version

```console
$ python --version
3.11.3
```

### aiohttp Version

```console
$ python -m pip show aiohttp
3.9.3
```

### multidict Version

```console
$ python -m pip show multidict
0
```

### yarl Version

```console
$ python -m pip show yarl
0
```

### OS

Linux

### Related component

Server

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

**Comment by Dreamsorcerer** [comment]
See https://docs.aiohttp.org/en/stable/web_advanced.html#graceful-shutdown

My suggestion is that long-lived tasks should be created independently, as per: https://docs.aiohttp.org/en/stable/web_advanced.html#complex-applications
Any task created before the application starts serving requests will be ignored from wait step and not hold the application open.

I'd use aiojobs for tasks that are expected to run for a short period of time in the background.

**Comment by novoxd** [comment]
I am wrong in the end of my last comment. It is happening because there is an `asyncio.Task(back_task)` running in background that were launched after aiohttp server start. On shutdown aiohttp will pick it and wait with 60s timeout for it to end, what will never happen. I thought that blocker is `iojobs._wait_failed` but because it is created in `cleanup_ctx` aiohttp is not waiting it.

I can say that it is an iojobs issue. But iohttp modification can help with this case and others. There is need for 3rd party packages to receive additional signal before aiohttp will start wait for background tasks to finish.

**Comment by Dreamsorcerer** [comment]
> In my case I have a background task that can run for 5-10 minutes and will be triggered by incoming request.

Yes, if this is started after the application is running, that won't work. I'm not clear what the task is doing, so I'm wondering if it doesn't make more sense to have a task started when the application starts up and then feed updates to it via a Queue or similar.

If not, maybe we can create a method to add tasks to the set of skipped tasks manually?

**Comment by novoxd** [comment]
I have a long running sh script launched by request. Will it be launched or not and amount of simultaneously running scripts is unknown from the start. Creating a bg task with equerries? Maybe. But it will be a duplication of logic what already iojobs contains.

In my case. From what I understood. Simplest solution from me is set timeout to zero. But it is a so so solution in wider case.

@Dreamsorcerer what do you think about using hypothetical three stage context in aiojobs and other 3d party packages? Like this. I think it will solve all type of this problems.
```py
def setup(app: web.Application, **kwargs: Any) -> None:
    async def cleanup_context(app: web.Application) -> AsyncIterator[None]:
        app[AIOJOBS_SCHEDULER] = scheduler = Scheduler(**kwargs)
        yield
        closing = asyncio.create_task(scheduler.close()) # runs after ^C
        yield
        await closing

    app.cleanup_3stage_ctx.append(cleanup_context)
```

**Comment by Dreamsorcerer** [comment]
Hmm, seems tricky as it wouldn't be backwards compatible (existing jobs would now cleanup before waiting on running handlers) and it would add complexity when most cases don't need it.

I'm still thinking we can add a method to put a task into the ignored list.

**Comment by Dreamsorcerer** [comment]
> Simplest solution from me is set timeout to zero.

Yes, in the short term, setting the timeout to something smaller would be the best approach (I'd still suggest giving it atleast 1 second though to give handlers a chance to send their responses).

**Comment by ralphm** [comment]
[This PR](https://github.com/aio-libs/aiohttp/pull/7718) reworked the shutdown procedure and I believe it is not entirely correct, either. Particularly on the definition of `runner.shutdown_callback` itself inside `_run_app`. This currently [reads](https://github.com/aio-libs/aiohttp/commit/107dd2445720cf354c84c3a271691f14e2e2875e#diff-dee5c5266d56f6e644698f884e7e75550501b4302d6b87229695ea01bd031637R343):

```python
    starting_tasks: "WeakSet[asyncio.Task[object]]" = WeakSet(asyncio.all_tasks())
    runner.shutdown_callback = partial(wait, starting_tasks, shutdown_timeout)
```

I have two questions here for @Dreamsorcerer:

1. At this point `asyncio.all_tasks()` will always include `_run_app` itself. It doesn't seem reasonable to wait on itself here. Should it not be excluded (by checking `asyncio.current_task(loop)`)?
2. Why does this not attempt to call `.cancel()` on the task, instead of waiting for the giving timeout? Long running tasks requiring a wait can still do this by shielding `CancelledError` (temporarily).

**Comment by Dreamsorcerer** [comment]
> [This PR](https://github.com/aio-libs/aiohttp/pull/7718) reworked the shutdown procedure and I believe it is not entirely correct, either. Particularly on the definition of `runner.shutdown_callback` itself inside `_run_app`. This currently [reads](https://github.com/aio-libs/aiohttp/commit/107dd2445720cf354c84c3a271691f14e2e2875e#diff-dee5c5266d56f6e644698f884e7e75550501b4302d6b87229695ea01bd031637R343):
> 
> ```python
>     starting_tasks: "WeakSet[asyncio.Task[object]]" = WeakSet(asyncio.all_tasks())
>     runner.shutdown_callback = partial(wait, starting_tasks, shutdown_timeout)
> ```
> 
> I have two questions here for @Dreamsorcerer:
> 
>     1. At this point `asyncio.all_tasks()` will always include `_run_app` itself.

No it doesn't (it'd be pretty obvious if it was wrong...):
https://github.com/aio-libs/aiohttp/commit/107dd2445720cf354c84c3a271691f14e2e2875e#diff-dee5c5266d56f6e644698f884e7e75550501b4302d6b87229695ea01bd031637R314-R316

>     2. Why does this not attempt to call `.cancel()` on the task, instead of waiting for the giving timeout? Long running tasks requiring a wait can still do this by shielding `CancelledError` (temporarily).

Again, refer to the documentation above. It gives time to allow currently running handlers to complete cleanly before cancelling them, thus avoiding users getting a dropped connection.

**Comment by Dreamsorcerer** [comment]
> 1. At this point `asyncio.all_tasks()` will always include `_run_app` itself. It doesn't seem reasonable to wait on itself here. Should it not be excluded (by checking `asyncio.current_task(loop)`)?

Wait, you've completely misread that code. That's a list of task to _not_ wait on...

**Comment by ralphm** [comment]
> Wait, you've completely misread that code. That's a list of tasks to _not_ wait on...

Gah, I see now that I got it wrong, and after a bit of digging now understand how it works. I have the same issue as @novoxd: my application can spawn long running tasks after the initial startup that I would like to be cancelled immediately upon graceful shutdown, so they can clean themselves up. Tasks created by aiojobs, e.g. by using `aiojobs.aiohttp.get_scheduler_from_app`, will unfortunately be waited on just the same.

**Comment by Dreamsorcerer** [comment]
Yes, they are still just tasks. Again, I'm thinking that a reasonable solution would be to create a method to add tasks to the exclude set, though happy to hear other ideas.

**Comment by Dreamsorcerer** [comment]
Just thinking over this again, this could be achieved without any changes to aiohttp, right? Because on_shutdown is triggered before waiting on handlers, we can just do something like:
```
async def on_startup(app):
    app["scheduler"] = Scheduler()

async def on_shutdown(app):
    await app["scheduler"].close()
```

**Comment by Dreamsorcerer** [comment]
We could something to the documentation for this..
